### PR TITLE
Datagen: move HashSet to the trait and other small cleanups

### PR DESCRIPTION
**Pull request by sffc**
Just some more things I had on my branch. No appreciable impact on performance (54.17, 58.70, 52.65).

### CODE REVIEW

File: provider/datagen/src/transform/icuexport/normalizer/mod.rs [hunk 1]
```diff
@@ -33,8 +33,8 @@ macro_rules! normalization_provider {
             }
         }
 
-        impl IterableDataProviderInternal<$marker> for crate::DatagenProvider {
-            fn supported_locales_impl(&self) -> Result<Vec<DataLocale>, DataError> {
+        impl IterableDataProvider<$marker> for crate::DatagenProvider {
+            fn supported_locales(&self) -> Result<Vec<DataLocale>, DataError> {
                 Ok(vec![DataLocale::default()])
             }
```
Comment (sffc) [inline provider/datagen/src/transform/icuexport/normalizer/mod.rs:39]: Note: I left the trivial impls directly with IterableDataProvider in the previous PR but accidentally migrated this trivial impl.

File: provider/datagen/src/transform/cldr/decimal/mod.rs [hunk 2]
```diff
@@ -72,7 +74,7 @@ impl crate::DatagenProvider {
             .collect())
     }
 
-    fn supported_locales(&self) -> Result<Vec<DataLocale>, DataError> {
+    fn supported_locales_for_numbers(&self) -> Result<HashSet<DataLocale>, DataError> {
```
Comment (sffc) [inline provider/datagen/src/transform/cldr/decimal/mod.rs:77]: This poorly named function has been around for a long time and I hope it didn't cause confusion. It dates to the time when each transformer was implemented on its own struct.

**Review by Manishearth — APPROVED** [review]

**Comment by robertbastian** [comment]
while you're in here please put the performance improvements in the changelog

**Comment by sffc** [comment]
Changelog in https://github.com/unicode-org/icu4x/pull/4484

**Pull request merged**